src/beamlime/constructors/providers.py

```diff
@@ -199,12 +203,54 @@ def __hash__(self) -> int:
     def __repr__(self) -> str:
         return f"Provider({self.call_name}, *{self.args}, **{self.keywords})."
 
+    @classmethod
+    def __copy__(cls, _obj: Provider[Product]) -> Provider[Product]:
+        return cls(_obj)
+
+    def __deepcopy__(self, _: dict[Any, Any]) -> Provider[Product]:
+        from copy import copy
+
+        return copy(self)
```

Comment:
This doesn't look like a deep copy. It needs to deepcopy all members. Doesn't the default deepcopy work here?

Reply:
It was mainly for making a deep copy of `_providers` dictionary of `ProviderGroup`.

I dug up the `deepcopy` and when you use `deepcopy` of a dictionary,
if the each value doesn't have a specific implementation of `__deepcopy__`,
it simply uses the same instance, not even `copy` of it...
```
from copy import deepcopy
provider = Provider(lambda: 0)
providers = {int: provider}
deepcopy(providers)  # means {deepcopy(int): deepcopy(provider)}
```

I could do something like
```
from copy import copy
new_providers = {key: copy(value) for key, value in providers.items()}
```
Rather than actually makes deep copy of every member of `Provider`....?

Reply:
Then I would call this `copy`, not `deepcopy`. If this is all you need, go ahead!

Reply:
I did...! Thank you!